### ISSUE TITLE
entropy prematurely plummeting

### ISSUE THREAD

**Issue by thegyro** [openai/universe-starter-agent#36]
I have been playing around with the starter code for quite some time. I wanted to tweak the architecture a bit and check if I can get the same results as the original A3C paper. Instead of an LSTM I am using just a FC layer after the convolutional layers. Instead of 42x42, I use 84x84 images. The _only_ changes I make to the existing code are following,

```
class FCPolicy(object):
    def __init__(self, ob_space, ac_space):
        self.x = x = tf.placeholder(tf.float32, [None] + list(ob_space))
        x = tf.nn.elu(conv2d(x, 16, "l{}".format(1), [8, 8], [4, 4], pad='VALID'))
        x = tf.nn.elu(conv2d(x, 32, "l{}".format(2), [4, 4], [2, 2], pad='VALID'))
        x = flatten(x)
        fc = tf.nn.elu(linear(x, 256, 'fc', normalized_columns_initializer(1.0)))
        self.logits = linear(fc, ac_space, "action", normalized_columns_initializer(0.01))
        self.vf = tf.reshape(linear(fc, 1, "value", normalized_columns_initializer(1.0)), [-1])
        self.sample = categorical_sample(self.logits, ac_space)[0, :]
        self.var_list = tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES,tf.get_variable_scope().name)

        self.dummy_features = [np.zeros(5), np.zeros(5)] #placeholders (required for LSTM but not for FC)

    def get_initial_features(self):
        return self.dummy_features
```

I am providing a wrapper which will give me four stacked frames in the following way

```
class AtariEnvWithHistory(gym.Wrapper):
    def __init__(self,env'):
        super(AtariEnvWithHistory, self).__init__(env)
        self.env = env
        shape = [84,84,4]
        self.observation_space = Box(0, 255, shape)

    def _reset(self):
        x_t = self.env.reset()
        self.s_t = np.stack((x_t,x_t,x_t,x_t), axis=2)
        return self.s_t

    def _step(self, action):
        x_t1, reward, terminal, info = self.env.step(action)
        s_t1 = np.append(self.s_t[:,:,1:], np.expand_dims(x_t1,axis=2),axis=2)
        self.s_t = s_t1
        return s_t1, reward, terminal, info
```

and process my frames

```
def _process_frame(frame):
    frame = frame[34:34+160, :160]
    frame = cv2.resize(frame, (84,84))
    frame = frame.mean(2)
    frame = frame.astype(np.float32)
    frame *= (1.0 / 255.0)
    frame = np.reshape(frame, [84,84])
    return frame
```
But the network converges to a very bad deterministic policy. The entropy quickly drops to zero. This was on PongDeterministic-v3
![entropy](https://cloud.githubusercontent.com/assets/4838833/21433691/0a442740-c871-11e6-8442-db98b0ce1611.png)

I think possible solutions can be to increase the entropy penalty or decrease the learning rate but I was just sticking to the parameters provided in the paper (I mean A3C is supposed to be more robust right?).  I wanted to know if I am doing something more fundamentally wrong.

**Comment by tlbtlbtlb** [comment]
Entropy collapse happens a lot. It's the default result if the model isn't capable of representing a useful policy, or the observations don't include the necessary info.

I'd be suspicious of the process_frame implementation. In tensorboard under the images tab, do the images look right?  The A3C object writes them with
```
            tf.image_summary("model/state", pi.x)
```
so you can check that they're cropped and resized correctly at the very end of the process.

To help debug, you can also throw in some calls like
```
   cv2.imshow('after-resize', frame)
   cv2.waitKey(0)
```
which will pop up a window with the image. Make sure you're not, for example, cropping off the right paddle.

**Comment by rafaljozefowicz** [comment]
Also worth noting than the current implementation tries to solve a more difficult problem than A3C in the paper (needed for real-time support). We have a separate thread used for collecting experience that runs in parallel to optimization. To make the setup more comparable, the thread should block and wait for the optimization to finish and update the parameters.

**Comment by thegyro** [comment]
@tlbtlbtlb Yes, I have tried looking at the images. They seem to be looking fine. This is what I get when I do cv2.imshow
![pong](https://cloud.githubusercontent.com/assets/4838833/21460091/dda42594-c944-11e6-8821-180693e3dad5.png)

And in tensorboard image summary,
![pong_tb](https://cloud.githubusercontent.com/assets/4838833/21460100/fc815d06-c944-11e6-82f4-ff8a08848acd.png)

Any idea what could be the potential problem?

**Comment by tlbtlbtlb** [comment]
One difference I notice between your code and the [A3C paper](https://arxiv.org/pdf/1602.01783.pdf) is that they use ReLU instead of ELU units.

**Comment by thegyro** [comment]
@tlbtlbtlb I have tried with ReLU as well. It doesn't seem to make any difference. I tried using the LSTM but with inputs as [84,84,4] and that seems to be working fine! Looks like the problem seems to be with using a FC layer instead of an LSTM.

@rafaljozefowicz  can this really affect performance that much?

**Comment by fgvbrt** [comment]
@thegyro I also tried to implement FC network instead of LSTM with 4 stacked frames 84x84 and did not get a good convergence. I also removed thread that collects training data and preforms actions in env, but still performance is very bad. Looks like it is much harder to train FC than LSTM policy.

**Comment by pengsun** [comment]
Hi @thegyro, have you solved the problem? I also observed this on some Atari games with my own tweak.

**Comment by theweaklink** [comment]
I don't think that would solve your problem but just to be consistent:
`self.observation_space = Box(0, 255, shape)` should be `self.observation_space = Box(0, 1, shape)` since you are converting the image from `uint8 [0-255]` to `float32 [0-1]` in your `_process_frame()` function.
